Redirect containers' stdout/stderr to /dev/null if log-driver is none

chenchun opened the pull request:
Currently, containers' stdout/stderr redirects to a pipe if log-driver
is none and no one will read from the it. Kernel will waste unnecessary
buffers to keep the data writen to the pipe. This is obviously wasteful

Signed-off-by: Chun Chen ramichen@tencent.com

LK4D4 commented on daemon/daemon.go:
```diff
@@ -193,9 +193,11 @@ func (daemon *Daemon) Register(container *Container) error {
-	container.stdout = broadcastwriter.New()
+	if container.getLogConfig().Type != "none" {
+		// Attach to stdout and stderr
+		container.stderr = broadcastwriter.New()
```
can't say that it's right, I might want to attach even with `--log-driver=none`

chenchun replied:
I want to redirect all the logs to /dev/null and I can't do that right now. Should I add another option `--log-driver=null` which redirect logs to `/dev/null`? 
I want to do this cause the oom issue of docker such as https://github.com/docker/docker/issues/14460
Or another option like `--log-driver=naive` which redirects logs to a simple file without passing through docker daemon?  
WDYT ?

LK4D4 commented:
It should be done other way I think. Now daemon just crashes if you try to run container with -it. It should be run option which work only with -d I think.

chenchun committed:
Redirect detach containers' stdout/stderr to /dev/null if log-driver=none

Currently, containers' stdout/stderr redirects to a pipe if log-driver
is none and no one will ever read from it. We will waste unnecessary kernel
buffers to keep the data writen to the pipe.

Signed-off-by: Chun Chen <ramichen@tencent.com>

LK4D4 commented:
Ok, now it works with -it, however attach just silently exit. We need to decide what we want.
ping @docker/core-maintainers

vdemeester commented:
@LK4D4 @chenchun We could return an error (or just a warning) if we try to attach to a container that its `log-driver` is pointing to `none`.

chenchun commented:
@LK4D4 @vdemeester thanks for the suggestion, rebased.
Found another problem during rebasing. Errors at  https://github.com/docker/docker/blob/master/api/server/router/local/container.go#L432 won't be properly shown by cli. Cli will display `Unrecognized input header` at https://github.com/docker/docker/blob/master/api/client/hijack.go#L218 instead of the real errors.

thaJeztah commented:
@chenchun sorry, looks like this needs another rebase; but we're still in design review.

thaJeztah commented:
ping @chenchun are you still working on this, or should we close this PR, are you awaiting feedback?

chenchun commented:
Sorry, I'm little busy these days, I'll rebase tomorrow. And I think design review doesn't depend on my rebase probably.

thaJeztah commented:
hm, looking back here, it looks like this PR will make `docker attach` not produce any output if `--log-driver=none`? I don't think that's expected behavior, i.e. I think users will expect being able to attach to a process, even if logging is disabled (because I don't see those two things as related to each other).

If that's the case, then I'm -1 for this change.

chenchun commented:
> make docker attach not produce any output

It actually throws an error "can't attach to a detached container whose log driver is none".
Close it since it won't be accepted.

thaJeztah commented:
Thanks for contributing @chenchun! Sorry this didn't get accepted
